File: Editorial/1196_How_Many_Apples_Can_You_Put_into_the_Basket/Python3/3_Counting_Sort.py
class Solution:
    def maxNumberOfApples(self, arr: List[int]) -> int:
        # initialize the counts to store all elements
        size = max(arr) + 1
        counts = [0] * size
        for weight in arr:
            counts[weight] += 1

        apples = units = 0
        for i in range(size):
            # if we have apples of i units of weight
            if counts[i] != 0:
                # we need to make sure that:
                # 1. we do not take more apples than those provided
                # 2. we do not exceed 5000 units of weight
                take = min(counts[i], (5000 - units) // i)
                if take == 0:
                    break

                apples += take
                units += take * i
        return apples

File: Editorial/1101-1200/1196_How_Many_Apples_Can_You_Put_into_the_Basket/Java/3_Counting_Sort.java
class Solution {
    public int maxNumberOfApples(int[] arr) {
        // initialize the counts to store all elements
        int size = -1;
        for (Integer weight: arr) {
            size = Math.max(size, weight);
        }
        int[] counts = new int[size + 1];
        for (Integer weight: arr) {
            counts[weight]++;
        }

        int apples = 0, units = 0;
        for (int i = 0; i < size + 1; i++) {
            // if we have apples of i units of weight
            if (counts[i] != 0) {
                // we need to make sure that:
                // 1. we do not take more apples than those provided
                // 2. we do not exceed 5000 units of weight
                int take = Math.min(counts[i], (5000 - units) / i);

                if (take == 0) {
                    break;
                }
                units += take * i;
                apples += take;

            }
        }
        return apples;
    }
}
